code/modules/shuttle/manipulator.dm

```diff
@@ -9,5 +9,214 @@
+
+	data["templates_tabs"] = sortList(data["templates_tabs"])
+
+	data["existing_shuttle"] = null
+
+	// Status panel
+	data["shuttles"] = list()
+	for(var/i in SSshuttle.mobile)
+		var/obj/docking_port/mobile/M = i
+		var/list/L = list()
+		L["name"] = M.name
+		L["id"] = M.id
+		L["timer"] = M.timer
+		L["timeleft"] = M.getTimerStr()
+		L["mode"] = capitalize(shuttlemode2str(M.mode))
+		L["status"] = M.getStatusText()
+		if(selected && selected.port_id == M.id)
+			existing_shuttle = M
+			data["existing_shuttle"] = L
+		data["shuttles"] += list(L)
+
+	// Modification panel
+	// should be disabled unless we are doing something?
+	// Show current proposal
+	// PREVIEW button to load
+	// CANCEL button to (unload if loaded) and return back
+	// LOAD button
+
+	// Multiple stage notifications
+	// - loaded
+	// - removed old shuttle
+	// - moved new shuttle to old location
+
+	//world.log << json_encode(data)
+	return data
+
+/obj/machinery/shuttle_manipulator/ui_act(action, params)
+	if(..())
+		return
+
+	var/mob/user = usr
+
+	switch(action)
+		if("select_template")
+			var/shuttle_id = params["shuttle_id"]
+
+			var/datum/map_template/shuttle/S = shuttle_templates[shuttle_id]
+			if(S)
+				src.selected = S
+				. = TRUE
+		if("jump_to")
+			if(params["type"] == "mobile")
+				for(var/i in SSshuttle.mobile)
+					var/obj/docking_port/mobile/M = i
+					if(M.id == params["id"])
+						user.loc = get_turf(M)
+		if("preview")
+			var/shuttle_id = params["id"]
+			var/datum/map_template/shuttle/S = shuttle_templates[shuttle_id]
+			if(S)
+				load_template(S)
+				if(preview_shuttle)
+					preview_shuttle_id = shuttle_id
+					user.loc = get_turf(preview_shuttle)
+		if("load")
+			var/shuttle_id = params["id"]
+			var/datum/map_template/shuttle/S = shuttle_templates[shuttle_id]
+			if(S)
+				if(!preview_shuttle || (S.shuttle_id != preview_shuttle_id))
+					if(preview_shuttle)
+						preview_shuttle.jumpToNullSpace()
+					load_template(S)
+					preview_shuttle_id = shuttle_id
+				// get the existing shuttle information, if any
+				var/timer
+				var/mode = SHUTTLE_IDLE
+				var/obj/docking_port/stationary/D
+				if(existing_shuttle)
+					timer = existing_shuttle.timer
+					mode = existing_shuttle.mode
+					D = existing_shuttle.get_docked()
+
+				if(!D)
+					var/msg1 = "[existing_shuttle] is not currently at a \
+						valid dock, the import will not continue."
+					WARNING(msg1)
+					message_admins(msg1)
+					preview_shuttle.jumpToNullSpace()
+					selected = null
+					return
+
+				// Destroy the old shuttle
+				existing_shuttle.jumpToNullSpace()
+				// Unflub the ID
+				var/new_id = replacetext(preview_shuttle.id, "(preview)", "")
+				preview_shuttle.id = new_id
+
+				preview_shuttle.dock(D)
+				preview_shuttle.timer = timer
+				preview_shuttle.mode = mode
+				. = TRUE
+
+				preview_shuttle = null
+				preview_shuttle_id = null
+				existing_shuttle = null
+				selected = null
+
+	update_icon()
```

Comment:
Pretty sure there is no icon? maybe only update if `. = TRUE` (ie UI needs updating)?